### PR TITLE
Move pyutilib.common exception references to pyomo.common.errors

### PR DESCRIPTION
**Pull request by mrmundt**
## Summary/Motivation:
To prepare for Pyomo's 6.0 release, we would like to do our best to move PyUtilib dependencies into Pyomo. This is the effort to move necessary pyutilib.common exceptions into pyomo.common.errors.

## Changes proposed in this PR:
- Add `pyutilib.common` code into `pyomo.common.errors`
- Update all references to items from `pyutilib.common`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/common/errors.py [hunk 1]
```diff
@@ -46,3 +47,27 @@ class InfeasibleConstraintException(PyomoException):
...
+    """
+
+    def __init__(self, *args, **kargs):
+        Exception.__init__(self, *args, **kargs)  #pragma:nocover
+
+
+if (sys.platform[0:3] != "win"):
+    WindowsError = WindowsError_def
+else:
+    WindowsError = WindowsError
```
Comment (jsiirola) [inline pyomo/common/errors.py:64]: I think we can remove this and update the two places in Pyomo that were catching `WindowsError` to just catch `OSError` (on Windows, the Python `WindowsError` is a subclass of `OSError`).

File: pyomo/common/errors.py [hunk 2]
```diff
@@ -46,3 +47,27 @@ class InfeasibleConstraintException(PyomoException):
...
+
+
+class ApplicationError(Exception):
```
Comment (jsiirola) [inline pyomo/common/errors.py:67]: Should this now be a subclass of `PyomoError`?

File: pyomo/common/errors.py [hunk 3]
```diff
@@ -46,3 +47,27 @@ class InfeasibleConstraintException(PyomoException):
...
+    """
+    An exception used when an external application generates an error.
+    """
+
+    def __init__(self, *args, **kargs):
+        Exception.__init__(self, *args, **kargs)  #pragma:nocover
```
Comment (jsiirola) [inline pyomo/common/errors.py:73]: This is nonfunctional and should just be:
```python
class ApplicationError(Exception):
    """
    An exception used when an external application generates an error.
    """
    pass
```

File: pyomo/dataportal/plugins/sheet.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ def open(self):
         else:
             try:
                 self.sheet = ExcelSpreadsheet(self.filename, ctype=self.ctype)
-            except pyutilib.common.ApplicationError:
+            except ApplicationError:
```
Comment (jsiirola) [inline pyomo/dataportal/plugins/sheet.py:57]: This is a problem: `ExcelSpreadsheet` is still going to raise a *pyutilib* `ApplicationError`, which will *not* be caught by a *pyomo* `ApplicationError`.

File: pyomo/dataportal/tests/test_dataportal.py [hunk 5]
```diff
@@ -65,7 +66,7 @@ def test_read_set(self):
             td.read()
             td.close()
             self.assertEqual( td._info, ['set', 'X', ':=', ('A1', 2.0, 3.0, 4.0), ('A5', 6.0, 7.0, 8.0), ('A9', 10.0, 11.0, 12.0), ('A13', 14.0, 15.0, 16.0)])
-        except pyutilib.common.ApplicationError:
+        except ApplicationError:
```
Comment (jsiirola) [inline pyomo/dataportal/tests/test_dataportal.py:69]: Do we know who is raising this `ApplicationError`?  Why are we allowing these exceptions to be "implicitly passed tests"?

File: pyomo/opt/tests/base/test_ampl.py [hunk 6]
```diff
@@ -52,7 +52,7 @@ def test3_write_nl(self):
         """ Convert from MOD+DAT to NL """
         try:
             self.model.write(currdir+'test3.nl')
-        except pyutilib.common.ApplicationError:
+        except ApplicationError:
```
Comment (jsiirola) [inline pyomo/opt/tests/base/test_ampl.py:55]: I am confused as to why we have these `except:` clauses.

**Review by jsiirola — COMMENTED** [review]
There is a lot of strange things going on here.  Do we have an idea on what the intent of the `ApplicationError` is?

**Comment by mrmundt** [comment]
> There is a lot of strange things going on here. Do we have an idea on what the intent of the `ApplicationError` is?

It's a lovely question. I'm not sure what the purpose is - but clearly, it's used in a lot of places. What do you think fits better?

File: pyomo/dataportal/plugins/sheet.py (hunk 4 above)
Comment (mrmundt) [reply pyomo/dataportal/plugins/sheet.py:57]: Ew. You're right. This is not meant to do anything to `pyutilib.excel.spreadsheet` so I will revert it for now.

File: pyomo/dataportal/tests/test_dataportal.py (hunk 5 above)
Comment (mrmundt) [reply pyomo/dataportal/tests/test_dataportal.py:69]: It looks like the DataManagerFactory functions (open, read, close) are raising this error - at least, that's what the code seems to imply. I'm not sure why that would raise an `ApplicationError`, though.

File: pyomo/opt/tests/base/test_ampl.py (hunk 6 above)
Comment (mrmundt) [reply pyomo/opt/tests/base/test_ampl.py:55]: I'm not sure either. What do you think is a better approach?

**Review by jsiirola — APPROVED** [review]

**Review by blnicho — APPROVED** [review]

**Pull request merged**